### PR TITLE
refactor(block): use BufferReader/BufferWriter

### PR DESCRIPTION
**Pull request by OttoAllmendinger**
The upstream PR that introduced BufferWriter has had some significant 
changes: https://github.com/bitcoinjs/bitcoinjs-lib/pull/1533

In particular, we now also have BufferReader. Also we can and should use
both classes in many more places.

Subclass BufferReader with `ZcashBufferReader` and implement
`readInt64()` there.

Replace fixtures for `readInt64()` with
`test/forks/zcash/bufferutils.js`.

Issue: BG-16466

### CODE REVIEW

File: src/bufferutils.js [hunk 1]
```diff
@@ -1,39 +1,141 @@
...
-  readInt64LE: readInt64LE,
-  writeUInt64LE: writeUInt64LE
+/**
+ * Helper class for serialization of bitcoin data types into a pre-allocated buffer.
+ */
+class BufferWriter {
+    constructor(buffer, offset = 0) {
+        this.buffer = buffer;
+        this.offset = offset;
+        typeforce(types.tuple(types.Buffer, types.UInt32), [buffer, offset]);
```
Comment (tylerlevine) [inline src/bufferutils.js]: How come this was moved after the assignments?

File: src/bufferutils.js [hunk 2]
```diff
@@ -1,39 +1,141 @@
...
-  b *= 0x100000000
-
-  verifuint(b + a, 0x001fffffffffffff)
-
-  return b + a
+function readUInt64LE(buffer, offset) {
+    const a = buffer.readUInt32LE(offset);
+    let b = buffer.readUInt32LE(offset + 4);
+    b *= 0x100000000;
+    verifuint(b + a, 0x001fffffffffffff);
```
Comment (tylerlevine) [inline src/bufferutils.js]: Where does this magic number come from?

Comment (tylerlevine) [reply src/bufferutils.js]: Oh, I see it in the tests, this is `Number.MAX_SAFE_INTEGER`.

File: src/bufferutils.js (hunk 1 above)
Comment (OttoAllmendinger) [reply src/bufferutils.js]: this is an artifact of the typescript compilation (originally `constructor(public buffer: Buffer, ...)`)

**Pull request merged**